rs/execution_environment/tests/canister_snapshots.rs

```diff
@@ -189,6 +191,169 @@ fn upload_snapshot_with_checkpoint() {
+        .unwrap();
+
+    env.execute_ingress(canister_id, "inc", vec![]).unwrap();
+    // take snapshot to learn valid metadata
+    let snapshot_id_orig = env
+        .take_canister_snapshot(TakeCanisterSnapshotArgs::new(canister_id, None))
+        .unwrap()
+        .snapshot_id();
+    let md_args = ReadCanisterSnapshotMetadataArgs::new(canister_id, snapshot_id_orig);
+    let md = env.read_canister_snapshot_metadata(&md_args).unwrap();
+    // download all data
+    let module_dl = env.get_snapshot_module(&md_args).unwrap();
+    assert_eq!(counter_canister_wasm, module_dl);
+    let heap_dl = env.get_snapshot_heap(&md_args).unwrap();
+    let stable_memory_dl = env.get_snapshot_stable_memory(&md_args).unwrap();
+
+    // load the snapshot with inconsistent hook status
+    load_faulty_snapshot(
+        &env,
+        Some(OnLowWasmMemoryHookStatus::Ready),
+        canister_id,
+        &md,
+        &module_dl,
+        &heap_dl,
+        &stable_memory_dl,
+    );
+    // load the snapshot with inconsistent hook status
+    load_faulty_snapshot(
+        &env,
+        Some(OnLowWasmMemoryHookStatus::Executed),
+        canister_id,
+        &md,
+        &module_dl,
+        &heap_dl,
+        &stable_memory_dl,
+    );
+}
+
+fn load_faulty_snapshot(
+    env: &StateMachine,
+    hook_status: Option<OnLowWasmMemoryHookStatus>,
+    canister_id: CanisterId,
+    md: &ReadCanisterSnapshotMetadataResponse,
+    module_dl: &[u8],
+    heap_dl: &[u8],
+    stable_memory_dl: &[u8],
+) {
+    // create a new snapshot via metadata upload
+    let args = UploadCanisterSnapshotMetadataArgs::new(
+        canister_id,
+        None,
+        module_dl.len() as u64,
+        md.exported_globals.clone(),
+        heap_dl.len() as u64,
+        stable_memory_dl.len() as u64,
+        md.certified_data.clone(),
+        None,
+        hook_status,
+    );
+    let snapshot_id = env
+        .upload_canister_snapshot_metadata(&args)
+        .unwrap()
+        .snapshot_id;
+    env.upload_snapshot_module(canister_id, snapshot_id.clone(), module_dl, None, None)
+        .unwrap();
+    env.upload_snapshot_heap(canister_id, snapshot_id.clone(), heap_dl, None, None)
+        .unwrap();
+    env.upload_snapshot_stable_memory(
+        canister_id,
+        snapshot_id.clone(),
+        stable_memory_dl,
+        None,
+        None,
+    )
+    .unwrap();
+    let _ = env.execute_ingress(canister_id, "inc", vec![]).unwrap();
+    let load_args = LoadCanisterSnapshotArgs::new(
+        canister_id,
+        SnapshotId::try_from(snapshot_id).unwrap(),
+        None,
+    );
+    let err = env.load_canister_snapshot(load_args).unwrap_err();
+    assert_eq!(err.code(), ErrorCode::InvalidManagementPayload);
+    assert!(err
+        .description()
+        .contains("uploaded snapshot is inconsistent with the canister's state"));
+}
+
+#[test]
+fn load_snapshot_inconsistent_metadata_globals_fails() {
```

Comment:
This test shares major parts with the previous one: taking and downloading snapshot, uploading snapshot data - it's just the snapshot metadata that are modified to become invalid (maybe a closure modifying the snapshot metadata would help unify the tests).